dm/pkg/checker/conn_checker.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checker
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/pingcap/tiflow/dm/dm/config"
+	"github.com/pingcap/tiflow/dm/pkg/conn"
+	tcontext "github.com/pingcap/tiflow/dm/pkg/context"
+	"github.com/pingcap/tiflow/dm/pkg/log"
+)
+
+type connAmountChecker struct {
+	targetDB *conn.BaseDB
+	stCfgs   []*config.SubTaskConfig
+
+	getConfigConn func(stCfgs []*config.SubTaskConfig) int
+	workerName    string
+}
+
+func newConnAmountChecker(targetDB *conn.BaseDB, stCfgs []*config.SubTaskConfig, fn func(stCfgs []*config.SubTaskConfig) int, workerName string) connAmountChecker {
+	return connAmountChecker{
+		targetDB:      targetDB,
+		stCfgs:        stCfgs,
+		getConfigConn: fn,
+		workerName:    workerName,
+	}
+}
+
+func (c *connAmountChecker) check(ctx context.Context, checkerName string) *Result {
+	result := &Result{
+		Name:  checkerName,
+		Desc:  "check if user-specified amount of connection exceeds downstream's maximum",
+		State: StateFailure,
+	}
+	baseConn, err := c.targetDB.GetBaseConn(ctx)
+	if err != nil {
+		markCheckError(result, err)
+		return result
+	}
+	var rows *sql.Rows
+	rows, err = baseConn.QuerySQL(tcontext.NewContext(ctx, log.L()), "SHOW GLOBAL VARIABLES LIKE 'max_connections'")
+	if err != nil {
+		markCheckError(result, err)
+		return result
+	}
+	defer rows.Close()
+	var (
+		maxConn   int
+		_variable string
+	)
+	for rows.Next() {
+		err = rows.Scan(&_variable, &maxConn)
+		if err != nil {
+			markCheckError(result, err)
+			return result
+		}
+	}
+	neededConn := c.getConfigConn(c.stCfgs)
+	if neededConn > maxConn {
+		result.Errors = append(
+			result.Errors,
+			NewError("downstream's max_connections: %d is less than the amount %s needs: %d", maxConn, c.workerName, neededConn),
+		)
+		result.Instruction = "set larger max_connections or reduce the pool size of dm"
```

Comment:
```suggestion
		result.Instruction = "set larger max_connections or adjust the configuration of dm"
```
I'm not sure if pool size is a common name for loader/syncer that user knows what we are referring :thinking:

Reply:
address in 0c04e1a

Reply:
https://github.com/pingcap/tiflow/commit/0c04e1a67e3ffaa25b4618d3d02ffa48267e9d63 is rollbacked